rsocket-core/src/main/java/io/rsocket/fragmentation/FrameFragmenter.java

```diff
@@ -24,168 +34,212 @@
 
-    if (state.hasReadableMetadata()) {
-      ByteBuf fragment = state.createFrame(byteBufAllocator, false, metadata, null);
-      logger.debug("Fragment {}", fragment);
-
-      sink.next(fragment);
-      return state;
+    ByteBuf metadataFragment = null;
+    if (metadata.isReadable()) {
+      // subtract the metadata frame length
+      remaining -= 3;
+      int r = Math.min(remaining, metadata.readableBytes());
+      remaining -= r;
+      metadataFragment = metadata.readRetainedSlice(r);
     }
 
-    ByteBuf data;
-    data = state.hasReadableData() ? state.readDataFragment(fragmentLength) : null;
-
-    if (state.hasReadableData()) {
-      ByteBuf fragment = state.createFrame(byteBufAllocator, false, metadata, data);
-      logger.debug("Fragment {}", fragment);
-
-      sink.next(fragment);
-      return state;
+    ByteBuf dataFragment = Unpooled.EMPTY_BUFFER;
+    if (remaining > 0 && data.isReadable()) {
+      int r = Math.min(remaining, data.readableBytes());
+      dataFragment = data.readRetainedSlice(r);
     }
 
-    ByteBuf fragment = state.createFrame(byteBufAllocator, true, metadata, data);
-    logger.debug("Final Fragment {}", fragment);
-
-    sink.next(fragment);
-    sink.complete();
-    return state;
-  }
-
-  private int getFragmentableLength(FragmentableFrame fragmentableFrame) {
-    return fragmentableFrame.getMetadataLength().orElse(0) + fragmentableFrame.getDataLength();
-  }
-
-  private boolean shouldFragment(ByteBuf frame) {
-    if (maxFragmentSize == 0 || !(frame instanceof FragmentableFrame)) {
-      return false;
+    switch (frameType) {
+      case REQUEST_FNF:
+        return RequestFireAndForgetFrameFlyweight.encode(
+            allocator, streamId, true, metadataFragment, dataFragment);
+      case REQUEST_STREAM:
+        return RequestStreamFrameFlyweight.encode(
+            allocator,
+            streamId,
+            true,
+            RequestStreamFrameFlyweight.initialRequestN(frame),
+            metadataFragment,
+            dataFragment);
+      case REQUEST_RESPONSE:
+        return RequestResponseFrameFlyweight.encode(
+            allocator, streamId, true, metadataFragment, dataFragment);
+      case REQUEST_CHANNEL:
+        return RequestChannelFrameFlyweight.encode(
+            allocator,
+            streamId,
+            true,
+            false,
+            RequestChannelFrameFlyweight.initialRequestN(frame),
+            metadataFragment,
+            dataFragment);
+        // Payload and synthetic types
+      case PAYLOAD:
+        return PayloadFrameFlyweight.encode(
+            allocator, streamId, true, false, false, metadataFragment, dataFragment);
+      case NEXT:
+        return PayloadFrameFlyweight.encode(
+            allocator, streamId, true, false, true, metadataFragment, dataFragment);
+      case NEXT_COMPLETE:
+        return PayloadFrameFlyweight.encode(
+            allocator, streamId, true, true, true, metadataFragment, dataFragment);
```

Comment:
nit: spec mentions `follows` flag should be ignored if `complete` flag is set - wondering if `complete` flag should be deferred to last fragment?

Reply:
not sure - the spec is a little unclear around fragmentation. If we send it last we can't generate the header until the last frame, and the metadata is supposed to come complete before the data - so how do you read it? Also I don't sent next_complete frames in the fragments i send next frames. Maybe we should change the spec because it seems more complicated to defer this till the end.